### DISCUSSION TITLE
Error compiling external Micropython module

### DISCUSSION THREAD

**Discussion by alescoppelli** [micropython/micropython#11795] [General]
Hi to all.
I am trying to compile the first example mentioned in the Micropython usermod documentation.
The error I get is as shown below.
Can anyone help me understand what is going on?

```
~/PICO_MICROPYTHON/micropython/ports/rp2$ make V=1 BOARD=PICO USER_C_MODULES=../../../moduli_per_pico  CFLAGS_EXTRA=-DMODULE_SIMPLEFUNCTION_ENABLED=1 all 
[ -e build-PICO/Makefile ] || cmake -S . -B build-PICO -DPICO_BUILD_DOCS=0 -DMICROPY_BOARD=PICO -DMICROPY_BOARD_DIR=/home/coppelli/PICO_MICROPYTHON/micropython/ports/rp2/boards/PICO -DUSER_C_MODULES=../../../moduli_per_pico
PICO_SDK_PATH is /home/coppelli/PICO_MICROPYTHON/micropython/lib/pico-sdk
Defaulting PICO_PLATFORM to rp2040 since not specified.
Defaulting PICO platform compiler to pico_arm_gcc since not specified.
PICO compiler is pico_arm_gcc
-- The C compiler identification is GNU 10.3.1
-- The CXX compiler identification is GNU 10.3.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/coppelli/ARM_COMPILERS/gcc-arm-none-eabi-10.3-2021.10/bin/arm-none-eabi-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/coppelli/ARM_COMPILERS/gcc-arm-none-eabi-10.3-2021.10/bin/arm-none-eabi-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The ASM compiler identification is GNU
-- Found assembler: /home/coppelli/ARM_COMPILERS/gcc-arm-none-eabi-10.3-2021.10/bin/arm-none-eabi-gcc
Build type is MinSizeRel
PICO target board is pico.
Using board configuration from /home/coppelli/PICO_MICROPYTHON/micropython/lib/pico-sdk/src/boards/include/boards/pico.h
-- Found Python3: /usr/bin/python3.8 (found version "3.8.10") found components: Interpreter 
TinyUSB available at /home/coppelli/PICO_MICROPYTHON/micropython/lib/tinyusb/src/portable/raspberrypi/rp2040; enabling build support for USB.
lwIP available at /home/coppelli/PICO_MICROPYTHON/micropython/lib/lwip
Including User C Module(s) from ../../../moduli_per_pico
CMake Error at /home/coppelli/PICO_MICROPYTHON/micropython/py/usermod.cmake:42 (include):
  include requested file is a directory:

    ../../../moduli_per_pico
Call Stack (most recent call first):
  CMakeLists.txt:78 (include)


Found User C Module(s): 
-- Configuring incomplete, errors occurred!
make: *** [Makefile:45: all] Error 1


```
At line 42 in usermod.cmake

```
# Include CMake files for user modules.
if (USER_C_MODULES)
    foreach(USER_C_MODULE_PATH ${USER_C_MODULES})
        message("Including User C Module(s) from ${USER_C_MODULE_PATH}")
        include(${USER_C_MODULE_PATH})
    endforeach()
endif()

```
Why there is the error message: "include requested file is a directory"?  In ""USER_C_MODULES" there are (there must be) the module's directory. 
Thanks to those who can help me.

**Comment by dlech** [comment]
The `USER_C_MODULES` variable should contain a list of files, not a directory.

**Reply by alescoppelli** [reply]
Dlexh, thank you your reply.

Reading the manual I thought that using  (in the command line) `USER_C_MODULES=../../../moduli_per_pico `  served to indicate   the root directory. Inside this (root) directory (**moduli_per_pico** )  (I thought) there should be there  other subdirectory.
Reading the manual, I think the indicated structure should be this:

* moduli_per_pico (root_dir)
    * samplefunction (module_folder)
        * samplefunction.c 
        * micropython.mk
   

From the manual it should be that `CFLAGS_EXTRA=-DMODULE_SIMPLEFUNCTION_ENABLED=1` is used to include or not include a module (i.e. the subfolder dedicated to that module; in this case simplefunction).

(1) Am I getting the command (`make`) parameters wrong?
or
(2) So how should the script be changed so that it traverses all subfolders and includes the files it finds there?

Considerations:

If (1) is true (I am in wrong) most probably the wrong manual (I did the same thing about a year ago and everything worked then)
If (2) is true (  the script wants only files not folder, how is it possible to structure a project with several modules (all files should be in one big folder)

**Reply by dlech** [reply]
I see what you mean. I wonder if the `py/usermod.cmake` file should be using `add_subdirectory(${USER_C_MODULE_PATH})` instead of `include(${USER_C_MODULE_PATH})`.

**Reply by jimmo** [reply]
@alescoppelli The documentaiton is a bit confusing because it describes both the `Make` and `CMake` process.

**For building the example modules which come with MicroPython,
set ``USER_C_MODULES``** to the ``examples/usercmodule`` directory for Make,
**or to ``examples/usercmodule/micropython.cmake`` for CMake.**

**Reply by alescoppelli** [reply]
Thank you very much Jimmo: the documentatin is ok. It's my fault: I should read the manual more carefully

**Comment by alescoppelli** [comment]
I immediately tried your suggestion but it still doesn't work:

`
CMake Error at /home/coppelli/PICO_MICROPYTHON/micropython/py/usermod.cmake:43 (add_subdirectory):
  add_subdirectory not given a binary directory but the given source
  directory "/home/coppelli/PICO_MICROPYTHON/moduli_per_pico" is not a
  subdirectory of "/home/coppelli/PICO_MICROPYTHON/micropython/ports/rp2".
  When specifying an out-of-tree source a binary directory must be explicitly
  specified.
Call Stack (most recent call first):
  CMakeLists.txt:80 (include)
`

Do you have any other suggestions?
